tests/utils.py

```diff
@@ -649,7 +649,7 @@ def run(self, result=None):
         for test_name in dir(self):
             if test_name.startswith('test_'):
                 expect_failure = hasattr(self, 'not_implemented') and test_name in self.not_implemented
-                getattr(self, test_name).__dict__['__unittest_expecting_failure__'] = expect_failure
+                getattr(self, test_name).__dict__.setdefault('__unittest_expecting_failure__', expect_failure)
```

Comment:
I think the fix you're looking for here is to correct the evaluation of expect_failure, rather than using setdefault on `__unittest_expecting_failure`. The implementation provided here would do; even that could be simplified to `expect_failure = getattr(self, 'not_implemented', False)`.